pkg/grpc/utils.go

```diff
@@ -35,42 +35,72 @@ var RPCTimeout = DefaultRPCTimeout
 
 var log = ctrl.Log.WithName("util")
 
-// CreateGrpcConnection create a grpc connection with given port and address
-func CreateGrpcConnection(address string, port int, caCertPath string, certPath string, keyPath string) (*grpc.ClientConn, error) {
-	if caCertPath != "" && certPath != "" && keyPath != "" {
-		var caCert, cert, key []byte
-		var err error
-		caCert, err = ioutil.ReadFile(caCertPath)
-		if err != nil {
-			return nil, err
-		}
-		cert, err = ioutil.ReadFile(certPath)
-		if err != nil {
-			return nil, err
-		}
-		key, err = ioutil.ReadFile(keyPath)
-		if err != nil {
-			return nil, err
-		}
-		return CreateGrpcConnectionFromRaw(address, port, caCert, cert, key)
+type TLSRaw struct {
+	CaCert []byte
+	Cert   []byte
+	Key    []byte
+}
+
+type TLSFile struct {
+	CaCert string
+	Cert   string
+	Key    string
+}
+
+type TLSFromType = string
+
+const (
+	RAW  TLSFromType = "RAW"
+	FILE TLSFromType = "FILE"
+)
+
+type CredentialProvider struct {
+	raw      TLSRaw
+	file     TLSFile
+	insecure bool
+	fromType TLSFromType
+}
+
+func (it *CredentialProvider) getCredentialOption() (grpc.DialOption, error) {
+	if it.insecure {
+		return grpc.WithInsecure(), nil
+	}
+	if it.fromType == RAW {
+		return it.TLSFromRaw()
 	}
-	options := []grpc.DialOption{grpc.WithUnaryInterceptor(TimeoutClientInterceptor)}
-	options = append(options, grpc.WithInsecure())
-	conn, err := grpc.Dial(fmt.Sprintf("%s:%d", address, port), options...)
+	if it.fromType == FILE {
+		return it.TLSFromFile()
+	}
```

Comment:
I prefer to use three implementations instead of if-else codes, and it is called "replace conditional with polymorphism". 

When more situations coming, we could just use appending without modifying to extend our application.

As the codes are not so complex, they're OK to me.

Reply:
Ahh I don't get it completely. Could you provide a small piece of example? `foobar` would make sense.

Reply:
Here is an example of that: https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html, this example uses a field that is an instance of `Swallow` instead of `bird.type`.

"Refactoring" only improves efficiency and maintenance, it will not change the usage of the origin thing.